File: assets/scripts/projecter.clj
#!/usr/bin/env bb

(ns projecter
  (:require [babashka.cli :as cli]
            [babashka.fs :as fs]
            [clojure.string :as str]))

(def allowed-templates #{"php" "python" "nodejs" "raw"})

(def cli-opts
  {:spec
   {:name {:desc "project name"
           :alias :n}

    :template {:desc "projects base template"
               :alias :t
               :validate allowed-templates
               :validate-msg (str "Template must be one of: "
                                  (str/join ", " allowed-templates))}

    :path {:desc "projects root"
           :alias :p}}})

(defrecord Template [^String name
                     ^String template
                     ^clojure.lang.IPersistentCollection tags])

(def template-path "/home/wmb/.local/templates/")
(def scripts-path "/home/wmb/.local/bin/")
(def scripts ["tunnel" ""])


(def templates
  {"php" (->Template "php"
                     "php.nix"
                     ["php" "laravel" "nodejs"])
   "python" (->Template "python"
                        "python.nix"
                        ["python" "uv" "pip"])
   "nodejs" (->Template "nodejs"
                        "nodejs.nix"
                        ["nodejs" "js" "ts"])
   "raw"   (->Template "raw"
                       "raw.nix"
                       ["raw" "agnostic"])})

(defn create-project
  [template path]
  (let [dir (fs/path (fs/cwd) path)]
    (when-not (fs/exists? dir)
      (fs/create-dirs dir))
    (fs/copy (fs/path template-path (:template template))
             (fs/path dir (:template template)))
    (doseq [i scripts] (fs/copy (fs/path (str scripts-path i "clj"))
                                (fs/path dir (:template template))))
    (println "New dev shell ready to be used")
    (println "Run: cd" dir)))


(defn create!
  [{:keys [template path]} templates]
  (let [tpl (get templates template)]
    (when-not tpl
      (throw (ex-info "Unknown template" {:template template})))
    (create-project tpl path)))

(defn main
  [& args]
  (let [opts (cli/parse-opts args cli-opts)]
    (create! opts templates)))

(apply main *command-line-args*)

File: assets/scripts/projecter.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'optparse'
require 'fileutils'

TEMPLATE_PATH = "#{Dir.home}/.local/templates/"
SCRIPTS_PATH  = "#{Dir.home}/.local/bin/"
SCRIPTS       = %w[tunnel].freeze

Template = Struct.new(:name, :template, :tags, keyword_init: true)

TEMPLATES = {
  "php"    => Template.new(name: "php",    template: "php.nix",    tags: %w[php laravel nodejs]),
  "python" => Template.new(name: "python", template: "python.nix", tags: %w[python uv pip]),
  "nodejs" => Template.new(name: "nodejs", template: "nodejs.nix", tags: %w[nodejs js ts]),
  "raw"    => Template.new(name: "raw",    template: "raw.nix",    tags: %w[raw agnostic]),
}.freeze

def create_project(template, path)
  dir = File.expand_path(path)
  FileUtils.mkdir_p(dir)
  FileUtils.cp(File.join(TEMPLATE_PATH, template.template), dir)
  SCRIPTS.each do |script|
    src = File.join(SCRIPTS_PATH, "#{script}.rb")
    FileUtils.cp(src, dir) if File.exist?(src)
  end
  puts "New dev shell ready to be used"
  puts "Run: cd #{dir}"
end

def parse_opts(argv)
  opts = {}
  OptionParser.new do |o|
    o.on("-n", "--name NAME",     "Project name")                                         { |v| opts[:name]     = v }
    o.on("-t", "--template TMPL", "Base template (#{TEMPLATES.keys.join(', ')})")         { |v| opts[:template] = v }
    o.on("-p", "--path PATH",     "Project root")                                         { |v| opts[:path]     = v }
  end.parse(argv)
  opts
end

def main(argv)
  opts = parse_opts(argv)
  abort("!! --path required") unless opts[:path]
  abort("!! --template required") unless opts[:template]

  tpl = TEMPLATES.fetch(opts[:template]) do
    abort("!! Unknown template: #{opts[:template]} — available: #{TEMPLATES.keys.join(', ')}")
  end

  create_project(tpl, opts[:path])
end

main(ARGV)
